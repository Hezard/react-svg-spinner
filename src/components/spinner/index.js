export { default as SpinnerWithPercentage, Spinner } from './spinner';
