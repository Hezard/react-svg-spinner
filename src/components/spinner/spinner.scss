@use 'sass:map';
@use "/scss/vars" as *;
@use "/scss/mixins" as *;
@use "/scss/animations" as *;

.spinner {
  position: relative;
  display: inline-block;
  overflow: hidden;
  transition: opacity 0.3s ease-in;

  &--primary & {
    &__circle--front {
      stroke: $primary;
    }
  }

  &--secondary & {
    &__circle--front {
      stroke: $secondary;
    }
  }

  &__svg {
    display: block;
  }

  &--disabled {
    opacity: 0.65;
  }

  &__svg--animation-rotate {
    animation: rotate 2s linear infinite paused;
  }

  &:not(&--disabled) & {
    &__svg {
      animation-play-state: running;
    }
  }

  &__circle {
    transform-origin: center;
    fill: transparent;

    &--front {
      stroke-linecap: round;
      transition: stroke-dashoffset 200ms linear, stroke 200ms linear;
    }

    &--back {
      stroke: $gray;
    }
  }
}

@each $key, $values in $spinner-sizes {
  .spinner--#{$key} {
    @include spinner-size(
      map.get($values, 'size'),
      map.get($values, 'stroke'),
      map.get($values, 'text')
    );
  }
}
