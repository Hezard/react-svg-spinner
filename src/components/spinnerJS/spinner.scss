@use "/scss/vars" as *;

.spinner {
  &--js & {
    &__circle--front {
      stroke-dasharray: var(--spinner-dash-array);
      stroke-dashoffset: var(--spinner-dash-offset);
    }
  }
}
