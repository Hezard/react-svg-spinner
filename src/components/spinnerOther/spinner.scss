@use "/scss/vars" as *;

.spinner {
  &--other & {
    &__circle--front {
      stroke-dasharray: var(--spinner-dash-array);
      stroke-dashoffset: var(--spinner-dash-offset);
    }
  }
}
