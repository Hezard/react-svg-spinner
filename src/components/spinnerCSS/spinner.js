import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withPercentage } from '../withPercentage';

import './spinner.scss';

export const SpinnerCSS = ({ children, className, color, disabled, mode, size, value }) => {
  const spinnerClass = classnames('spinner', className, `spinner--${size}`, `spinner--${color}`, {
    'spinner--disabled': disabled
  });
  const spinnerSvgClass = classnames('spinner__svg', `spinner__svg--animation-${mode}`);

  const style = {
    '--spinner-dash-offset': value
  };

  return (
    <div className={spinnerClass}>
      <svg className={spinnerSvgClass} shapeRendering="geometricPrecision">
        <circle
          className="spinner__circle spinner__circle--back"
          r="50%"
          cx="50%"
          cy="50%"
        ></circle>
        <circle
          className="spinner__circle spinner__circle--front"
          style={style}
          r="50%"
          cx="50%"
          cy="50%"
        ></circle>
      </svg>
      {children}
    </div>
  );
};

SpinnerCSS.defaultProps = {
  color: 'primary',
  disabled: false,
  mode: 'rotate',
  size: 'md',
  value: 0
};

SpinnerCSS.propTypes = {
  className: PropTypes.string,
  color: PropTypes.oneOf(['primary', 'secondary']),
  disabled: PropTypes.bool,
  mode: PropTypes.oneOf(['rotate']),
  size: PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
  value: PropTypes.number
};

export default withPercentage(SpinnerCSS);
