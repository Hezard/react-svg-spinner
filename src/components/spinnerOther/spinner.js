import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { withPercentage } from '../withPercentage';

import './spinner.scss';

export const SpinnerOther = ({
  circumference,
  children,
  className,
  color,
  disabled,
  mode,
  stroke,
  value
}) => {
  const spinnerClass = classnames('spinner', className, 'spinner--other', `spinner--${color}`, {
    'spinner--disabled': disabled
  });
  const spinnerSvgClass = classnames('spinner__svg', `spinner__svg--animation-${mode}`);

  const radius = circumference / 2 - stroke / 2;
  const path = 2 * Math.PI * radius;
  const linecapGap = 1; // linecap fix for Edge and Firefox
  const style = {
    '--spinner-dash-offset': ((path + linecapGap / 2) * (100 - value)) / 100,
    '--spinner-dash-array': `${path} ${path + linecapGap}`
  };

  return (
    <div className={spinnerClass}>
      <svg
        width={circumference}
        height={circumference}
        className={spinnerSvgClass}
        shapeRendering="geometricPrecision"
        data-testid="spinner-svg"
      >
        <circle
          className="spinner__circle spinner__circle--back"
          r={radius}
          cx="50%"
          cy="50%"
          strokeWidth={stroke}
          data-testid="spinner-circle-back"
        ></circle>
        <circle
          className="spinner__circle spinner__circle--front"
          style={style}
          r={radius}
          cx="50%"
          cy="50%"
          strokeWidth={stroke}
          data-testid="spinner-circle-front"
        ></circle>
      </svg>
      {children}
    </div>
  );
};

SpinnerOther.defaultProps = {
  color: 'primary',
  disabled: false,
  mode: 'rotate',
  circumference: 170,
  stroke: 10
};

SpinnerOther.propTypes = {
  className: PropTypes.string,
  color: PropTypes.oneOf(['primary', 'secondary']),
  circumference: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  disabled: PropTypes.bool,
  mode: PropTypes.oneOf(['rotate']),
  stroke: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  value: PropTypes.number
};

export default withPercentage(SpinnerOther);
