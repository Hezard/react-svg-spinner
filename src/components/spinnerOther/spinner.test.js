import React from 'react';
import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { SpinnerOther } from './spinner';

afterEach(cleanup);

test('SpinnerOther with javascript', () => {
  const props = {
    color: 'primary',
    className: 'spinner-test',
    size: 'lg'
  };
  const { container, rerender, getByTestId } = render(<SpinnerOther {...props}></SpinnerOther>);

  expect(container.firstChild).toMatchSnapshot();

  rerender(<SpinnerOther {...props} disabled={true}></SpinnerOther>);

  expect(container.firstChild).toHaveClass('spinner--disabled');

  rerender(<SpinnerOther {...props} circumference="120" stroke="6"></SpinnerOther>);

  expect(getByTestId('spinner-svg')).toHaveAttribute('width', '120');
  expect(getByTestId('spinner-svg')).toHaveAttribute('height', '120');
  expect(getByTestId('spinner-circle-back')).toHaveAttribute('r', '57');
  expect(getByTestId('spinner-circle-front')).toHaveAttribute('r', '57');
});
