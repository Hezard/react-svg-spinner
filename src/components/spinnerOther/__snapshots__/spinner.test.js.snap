// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerOther with javascript 1`] = `
<div
  class="spinner spinner-test spinner--other spinner--primary"
>
  <svg
    class="spinner__svg spinner__svg--animation-rotate"
    data-testid="spinner-svg"
    height="170"
    shape-rendering="geometricPrecision"
    width="170"
  >
    <circle
      class="spinner__circle spinner__circle--back"
      cx="50%"
      cy="50%"
      data-testid="spinner-circle-back"
      r="80"
      stroke-width="10"
    />
    <circle
      class="spinner__circle spinner__circle--front"
      cx="50%"
      cy="50%"
      data-testid="spinner-circle-front"
      r="80"
      stroke-width="10"
    />
  </svg>
</div>
`;
